### PR TITLE
[OBSOLETE] Added Sign Language Transformer.

### PR DESCRIPTION
**Pull request by wiktorfilipiuk**

### CODE REVIEW

**Review by adrianlachowicz — APPROVED** [review]

**Comment by majsylw** [comment]
@wiktorfilipiuk 
File "train.py", line 148, in main
    model = GlossTranslationModel(
TypeError: __init__() got an unexpected keyword argument 'num_segments'

**Review by majsylw — COMMENTED** [review]
@wiktorfilipiuk it works at this branch now, please rebase with current main and change this PR to push code there

File: models/transformers/sign_language_transformer.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    def forward(self, input: torch.Tensor):
+        # SLRT Start
+        positional_encoding_normalized = self._positional_encoding_norm(input)
+
+        # Self Attention (Multi-Head Attention) Start
+        values = positional_encoding_normalized
+        keys = positional_encoding_normalized
+        questions = positional_encoding_normalized
+
+        attention_output = MultiHeadedAttention(num_heads=self._num_segments,
```
Comment (AgaMiko) [inline models/transformers/sign_language_transformer.py:108]: `MultiHeadedAttention` should be defined in constructor (`__init__`)

File: requirements.txt [hunk 2]
```diff
@@ -5,4 +5,5 @@ torch-optimizer~=0.3.0
 timm~=0.4.12
 scikit-learn~=0.24.2
 neptune-client==0.14.1
-transformers==4.15.0
\ No newline at end of file
+transformers==4.15.0
+transformers==4.15.0
```
Comment (AgaMiko) [inline requirements.txt:9]: Same lib two times

File: models/transformers/sign_language_transformer.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                                                       feedforward_size=feedforward_size,
+                                                       num_segments=num_segments,
+                                                       dropout_rate=dropout_rate)])
+        self._dropout_positional_encoding = nn.Dropout(dropout_rate)
+        self._last_norm = nn.LayerNorm(input_size)
+        self._last_linear = nn.Linear(num_segments*input_size, output_size)
+
+    def forward(self, input: torch.Tensor):
+        # Positional Encoding Start
+        max_len = 10000
```
Comment (AgaMiko) [inline models/transformers/sign_language_transformer.py:45]: should be moved to `__init__`

File: models/transformers/sign_language_transformer.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+    def __init__(self,
+                 num_heads: int,
+                 size: int,
+                 dropout_rate: float = 0.1):
+        """
+        Layer implemented as presented in the paper "Attention is all you need"
+        https://arxiv.org/pdf/1706.03762.pdf
+        Args:
+            num_heads (int): Number of input images per single video.
```
Comment (polquant) [inline models/transformers/sign_language_transformer.py:140]: Generally one can have different (usually much lower) number of heads than images. If you have 100 frames per sequence, 100 heads is a lot.

File: models/transformers/sign_language_transformer.py (hunk 1 above)
Comment (polquant) [inline models/transformers/sign_language_transformer.py:108]: This will be a lot of heads

File: models/transformers/sign_language_transformer.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        # Positional Encoding Start
+        max_len = 10000
+        position_encoding = torch.zeros(max_len, self._input_size)
+        position = torch.arange(0, max_len).unsqueeze(1)
+        div_term = torch.exp(
+            (torch.arange(0, self._input_size, 2, dtype=torch.float) * -(math.log(10000.0) / self._input_size))
+        )
+        position_encoding[:, 0::2] = torch.sin(position.float() * div_term)
+        position_encoding[:, 1::2] = torch.cos(position.float() * div_term)
+        position_encoding = position_encoding.unsqueeze(0)
```
Comment (polquant) [inline models/transformers/sign_language_transformer.py:53]: `position_encoding` ->` self._position_encoding` in `__init__`

File: models/transformers/sign_language_transformer.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+        positional_encoding = input + position_encoding[:, :input.shape[1]]
```
Comment (polquant) [inline models/transformers/sign_language_transformer.py:55]: `position_encoding` -> `self._position_encoding`

File: models/transformers/sign_language_transformer.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+            feedforward_size (int): Number of features in intermediate feedforward "layer".
+            num_encoder_layers (int): Number of encoder layers.
+            num_segments (int): Number of frames in a video.
+            dropout_rate (float): Dropout rate.
+        """
+        super(SignLanguageTransformer, self).__init__()
+        self._dropout_rate = dropout_rate
+        self._feedforward_size = feedforward_size
+        self._input_size = input_size
+        self._num_encoder_layers = num_encoder_layers
```
Comment (polquant) [inline models/transformers/sign_language_transformer.py:33]: is it used further?

File: models/transformers/sign_language_transformer.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        # we add a dimension for the heads to it below: [B, 1, 1, M]
+        if mask is not None:
+            scores = scores.masked_fill(~mask.unsqueeze(1), float("-inf"))
+
+        # apply attention dropout and compute context vectors.
+        attention = self.softmax(scores)
+        attention = self._dropout(attention)
+
+        # get context vector (select values with attention) and reshape
+        # back to [B, M, D]
```
Comment (polquant) [inline models/transformers/sign_language_transformer.py:194]: What is `[B, M D]`? ;)

File: models/transformers/sign_language_transformer.py (hunk 3 above)
Comment (wiktorfilipiuk) [reply models/transformers/sign_language_transformer.py:56]: From what I understood this value is not really any mutable parameter. I don't expect it to be changed in any way.

Comment (polquant) [reply models/transformers/sign_language_transformer.py:56]: So it is a class variable (static), and should be put between clas... and def __init___

File: models/transformers/sign_language_transformer.py (hunk 8 above)
Comment (wiktorfilipiuk) [reply models/transformers/sign_language_transformer.py:222]: B - number of items in a batch
M - number of segments * number of frames per segment
D - number of features

Comment (polquant) [reply models/transformers/sign_language_transformer.py:222]: I think it should be written in the commented line :)

Comment (wiktorfilipiuk) [reply models/transformers/sign_language_transformer.py:222]: Done :D

**Comment by wiktorfilipiuk** [comment]
I leave this PR for the time being to keep the comments in mind.

**Pull request closed without merging**